multicluster/controllers/multicluster/node_controller.go

```diff
@@ -0,0 +1,202 @@
+)
+
+type (
+	// NodeReconciler is for member cluster only.
+	// It will create a Gateway object if a Node has an annotation `multicluster.antrea.io/gateway:true`
+	// and update corresponding Gateway if any subnets changes.
+	NodeReconciler struct {
+		client.Client
+		Scheme           *runtime.Scheme
+		namespace        string
+		installedGWNodes cache.Indexer
+		precedence       mcsv1alpha1.Precedence
+	}
+)
+
+// NodeReconciler will watch Node object change and create a
+// corresponding Gateway if the Node has an annotation
+// `multicluster.antrea.io/gateway:true`.
+func NewNodeReconciler(
+	Client client.Client,
+	Scheme *runtime.Scheme,
+	namespace string,
+	precedence mcsv1alpha1.Precedence) *NodeReconciler {
+	reconciler := &NodeReconciler{
+		Client:    Client,
+		Scheme:    Scheme,
+		namespace: namespace,
+		installedGWNodes: cache.NewIndexer(gatewayNodeKeyFunc, cache.Indexers{
+			"ips": gatewayNodeIndexerByIPsFunc,
+		}),
+		precedence: precedence,
+	}
+	return reconciler
+}
+
+func gatewayNodeKeyFunc(obj interface{}) (string, error) {
+	gwe := obj.(*mcsv1alpha1.Gateway)
+	return gwe.Namespace + gwe.Name, nil
+}
+
+func gatewayNodeIndexerByIPsFunc(obj interface{}) ([]string, error) {
+	gwe := obj.(*mcsv1alpha1.Gateway)
+	return []string{gwe.InternalIP, gwe.GatewayIP}, nil
+}
+
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways/finalizers,verbs=update
+
+func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("reconciling Node", "node", req.Name)
+	if string(r.precedence) == "" {
+		r.precedence = mcsv1alpha1.PrecedencePrivate
+	}
+
+	gwNode := &mcsv1alpha1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: r.namespace,
+		},
+	}
+	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
+	//   - Delete the Gateway if the Node is deleted
+	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
+	//   - Create a new Gateway if there is no existing Gateway
+	node := &corev1.Node{}
+	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := r.Client.Delete(ctx, gwNode, &client.DeleteOptions{}); err != nil {
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+			return ctrl.Result{}, nil
+		}
+		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	_, isGWNode := node.Annotations[common.GatewayAnnotation]
+	var err error
+	var gwIP, internalIP string
+	if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
+		return ctrl.Result{}, err
```

Comment:
If no valid IP, we should not retry, but should wait for the Node update.

Reply:
done